### PR TITLE
adding support for newline

### PR DESCRIPTION
**Pull request by fatemeh-affinity**
This PR adds support for newline `<br>' and paragraph `<p>` to slack transformer.

Newline conversion:
`<br>newline`: "\nnewline"`

Paragraph conversion:
`<p>Hello World</p>` : `"Hello World"`
`Hello<p>World</p>` : `"Hello\nWorld"`
`<p>Hello</p>World` : `"Hello\nWorld"`
`Hello<p></p>World` : `"Hello\nWorld"`
`Hello<p></p> <p></p> <p></p>World` : `"Hello\nWorld"`
`Hello World <p></p> ` : `"Hello World\n"`

Please take a look at the tests for more clarity.

### CODE REVIEW

File: lib/slack_transformer/html/newline.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        @input = input
+      end
+
+      def to_slack        
+        fragment = Nokogiri::HTML.fragment(input)
+
+        fragment.children.each do |child|
+          case child.name
+          when 'p'
+            child.replace("\n #{child.children.to_html} \n")
```
Comment (arjunsr1) [inline lib/slack_transformer/html/newline.rb]: Do we want it wrapped in newlines, or just a newline at the end before the next tag to come along?

**Review by arjunsr1 — DISMISSED** [review]
1 question

File: lib/slack_transformer/html/newline.rb (hunk 1 above)
Comment (fatemeh-affinity) [reply lib/slack_transformer/html/newline.rb]: This is how they appear:
image

I guess I should replace the line with this:
`child.replace("\n\n #{child.children.to_html} \n\n")`

Comment (fatemeh-affinity) [reply lib/slack_transformer/html/newline.rb]: image

Comment (mimslee32) [reply lib/slack_transformer/html/newline.rb]: hm i'm not sure i'm fully following your screenshots. i don't think that we should be adding in a newline before the `<p>` tag. And we may not need a newline at the end if there are no more elements after it 
so: `<p>Hi</p>` should just become `Hi`, but `Hello<p>World</p>` should become
```
Hello
World
```
And `Hello<p>World</p>Again` should become
```
Hello
World
Again
```

**Review by mimslee32 — CHANGES_REQUESTED** [review]
Should not be adding in newline before and after p tag everytime

File: lib/slack_transformer/html.rb [hunk 2]
```diff
@@ -18,7 +20,9 @@ class Html
       SlackTransformer::Html::Strikethrough,
       SlackTransformer::Html::Code,
       SlackTransformer::Html::Preformatted,
-      SlackTransformer::Html::Hyperlinks
+      SlackTransformer::Html::Hyperlinks,
```
Comment (mimslee32) [inline lib/slack_transformer/html.rb]: Hyperlinks actually has to be done at the very end, and any transformer using nokogiri should be done first. This is because as we transform the html string to mrkdown, it becomes invalid html!

File: lib/slack_transformer/html/paragraph.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        @input = input
+      end
+
+      def to_slack
+        fragment = Nokogiri::HTML.fragment(input)
+        previous = nil
+        
+        fragment.children.each do |child|
+          if child.name == 'p' 
+
```
Comment (mimslee32) [inline lib/slack_transformer/html/paragraph.rb]: remove unnecessary newline :p

File: lib/slack_transformer/html/paragraph.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        previous = nil
+        
+        fragment.children.each do |child|
+          if child.name == 'p' 
+            
+            if previous.nil?
+              child.replace("#{child.children.to_html}")
+            else
+              child.replace("\n#{child.children.to_html}")
+            end
```
Comment (mimslee32) [inline lib/slack_transformer/html/paragraph.rb]:
```suggestion
            newline = previous.nil? ? "" : "\n"
            child.replace("#{newline}#{child.children.to_html}")
```

File: lib/slack_transformer/html/paragraph.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      def initialize(input)
+        @input = input
+      end
+
+      def to_slack
+        fragment = Nokogiri::HTML.fragment(input)
+        previous = nil
+        
+        fragment.children.each do |child|
+          if child.name == 'p'
```
Comment (arjunsr1) [inline lib/slack_transformer/html/paragraph.rb]: Can you make a constant at the top `P_TAG = 'p'`

Then have something like this:
```
tag_type = child.name
if tag_type == P_TAG
    ...
    previous = child.children.empty? ? nil : tag_type
else
    child.replace("\n#{child.to_html}") if previous == P_TAG            
    previous = tag_type
```

File: lib/slack_transformer/html/paragraph.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+          if child.name == 'p' 
+            
+            if previous.nil?
+              child.replace("#{child.children.to_html}")
+            else
+              child.replace("\n#{child.children.to_html}")
+            end
+
+            previous = child.children.empty? ? nil : 'p'
+
```
Comment (mimslee32) [inline lib/slack_transformer/html/paragraph.rb]: remove newline

File: lib/slack_transformer/html/paragraph.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        fragment.children.each do |child|
+          if child.name == 'p' 
+            
+            if previous.nil?
+              child.replace("#{child.children.to_html}")
+            else
+              child.replace("\n#{child.children.to_html}")
+            end
+
+            previous = child.children.empty? ? nil : 'p'
```
Comment (mimslee32) [inline lib/slack_transformer/html/paragraph.rb:20]: i'm not sure i'm fully following this line :o can you explain it a bit to me?

Comment (mimslee32) [reply lib/slack_transformer/html/paragraph.rb:20]: Oh looking at the tests, i see what you mean. Maybe a quick comment here to denote that this is a special edge case where the p tag is empty

**Review by mimslee32 — APPROVED** [review]

File: lib/slack_transformer/html.rb (hunk 2 above)
Comment (fatemeh-affinity) [reply lib/slack_transformer/html.rb]: Moved newline and paragraph at the very beginning.
Just wondering, if Hyperlink at the very end is safe considering other transformer, using gsub, might also make the HTML invalid, or not?

Comment (fatemeh-affinity) [reply lib/slack_transformer/html.rb]: Just noticed [this comment](https://github.com/affinity/slack_transformer/blob/c600241a55e3c41cdfdc6d81c8231003ac26231f/lib/slack_transformer/html.rb#L16) of yours, it's related to my question.

File: lib/slack_transformer/html/paragraph.rb (hunk 5 above)
Comment (fatemeh-affinity) [reply lib/slack_transformer/html/paragraph.rb]: Let's do that in another PR if we want to go with this approach, using constants, since all the other newly added transformers, list, hyperlink and newline should be refactored to be consistent.

File: lib/slack_transformer/html/paragraph.rb (hunk 7 above)
Comment (fatemeh-affinity) [reply lib/slack_transformer/html/paragraph.rb:20]: Done,

File: lib/slack_transformer/html/paragraph.rb (hunk 3 above)
Comment (fatemeh-affinity) [reply lib/slack_transformer/html/paragraph.rb]: Done,
